tests/camera.cpp

```diff
@@ -0,0 +1,177 @@
+/** @file
+ *  Program entry point.
+ *
+ *  @copyright 2006  Palm Studios
+ *
+ *  @author Jesse Buhagiar
+ */
+
+#include "SH3/arc/mft.hpp"
+#include "SH3/system/config.hpp"
+#include "SH3/system/exit_code.hpp"
+#include "SH3/system/log.hpp"
+#include "SH3/system/window.hpp"
+#include "SH3/system/glprogram.hpp"
+#include "SH3/system/glbuffer.hpp"
+#include "SH3/system/glvertarray.hpp"
+#include "SDL2/SDL.h"
+#include "SH3/camera/camera.hpp"
+#include <cstdio>
+
+static GLfloat g_vertex_buffer_data[] = {
+   -1.0f, -1.0f, 0.0f,
+   1.0f, -1.0f, 0.0f,
+   0.0f,  1.0f, 0.0f,
+};
+
+static GLfloat glCol[] = {
+    1.0f, 0.0f, 0.0f,
+    0.0f, 1.0f, 0.0f,
+    0.0f, 0.0f, 1.0f,
+};
+
+/**
+ *  VAO Attributes for a colorful triangle.
+ */
+struct TriangleAttributes final
+{
+private:
+    using Target = sh3_gl::buffer_object::Target;
+
+    TriangleAttributes() = delete;
+
+public:
+    enum class Slot
+    {
+        VERTEX,
+        COLOR,
+        MAX
+    };
+
+    static constexpr sh3_gl::vao_target_array<Slot> Targets =
+    {
+        Target::ARRAY_BUFFER,
+        Target::ARRAY_BUFFER
+    };
+};
+constexpr sh3_gl::vao_target_array<TriangleAttributes::Slot> TriangleAttributes::Targets;
+
+/**
+ *  Entry point to the program.
+ *
+ *  @note Requires @c argc and @c argv to make it compatible with SDL2.
+ *
+ *  @param argc Number of arguments.
+ *  @param argv Argument vector.
+ *
+ *  @returns An @ref exit_code specifying success or failure.
+ */
+int main(int argc, char** argv)
+{
+    static_cast<void>(argc);
+    static_cast<void>(argv);
+
+
+    Log(LogLevel::INFO, "===SILENT HILL 3 REDUX===");
+    Log(LogLevel::INFO, "Copyright 2016-2017 Palm Studios\n");
+
+    sh3_window window(640, 480, "sh3redux");
+
+    bool quit = false;
+    SDL_Event e;
+
+    sh3_gl::program::load_error err;
+    sh3_gl::program test("test", err);
+
+    using Triangle = sh3_gl::vao<TriangleAttributes>;
+
+    Triangle triVao;
+
+    triVao.Bind();
+
+    angle<float> fov;
+
+    sh3_gl::buffer_object& verts = triVao[Triangle::Slot::VERTEX];
+    verts.BufferData(g_vertex_buffer_data, sizeof(g_vertex_buffer_data), GL_STATIC_DRAW);
+    triVao.SetDataLocation(Triangle::Slot::VERTEX, Triangle::DataType::FLOAT, 3, 0, 0);
+
+    sh3_gl::buffer_object& col = triVao[Triangle::Slot::COLOR];
+    col.BufferData(glCol, sizeof(glCol), GL_STATIC_DRAW);
+    triVao.SetDataLocation(Triangle::Slot::COLOR, Triangle::DataType::FLOAT, 3, 0, 0);
+
+    triVao.Unbind();
+
+    sh3::camera::Camera cam(glm::vec3(0,0,0), fov.FromDegrees(60), (float)640/(float)480, 0.001f, 100.0f);
+    cam.SetPosition(glm::vec3(3, 0, 4));
+    cam.LookAt(glm::vec3(0, 0, 0));
+    cam.SetMode(sh3::camera::MODE::FIRST_PERSON);
+
+    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+
+    GLuint mID = glGetUniformLocation(test.GetID(), "mvp");
+    glm::mat4 model = glm::mat4(1.0f);
+    float x = 0.0f;
+
+    bool forward = false;
+    bool backward = false;
+
+    angle<float> pitch;
+    angle<float> yaw;
```

Comment:
I thought the camera was broken for me, but it turns out that `yaw` is initialized with 0, but the camera is initialized with a different value.
Either the camera should get a `AddYaw` or a `GetYaw` function so we don't need to keep track of the angles ourselves (same with `pitch`).

Reply:
`cam.Pitch()` is supposed to this, but I was using the function entirely wrong...